apps/music/js/queue.js

```diff
@@ -0,0 +1,335 @@
+   */
+  function BaseQueue() {
+    this._index = null;
+  }
+
+  BaseQueue.prototype = {
+    /**
+     * Initialize the playback queue.
+     *
+     * @param {Number} [index] The index of the queue to start at.
+     */
+    _init: function(index) {
+      if (this.length === 0) {
+        throw Error('cannot have an empty queue');
+      }
+      this._index = index || 0;
+      this._shuffle(playbackSettings.shuffle, index !== undefined);
+    },
+
+    /**
+     * Shuffle or unshuffle the playback queue. If the queue is already
+     * shuffled, this will reshuffle it.
+     *
+     * @param {Boolean} enabled True if the queue should be shuffled, false
+     *        otherwise.
+     * @param {Boolean} keepIndex True if the current queue index should be
+     *        retained, false otherwise.
+     */
+    _shuffle: function(enabled, keepIndex = true) {
+      this._shuffleTime = Date.now();
+      if (!enabled) {
+        if (keepIndex) {
+          this._index = this.index;
+        }
+        delete this._shuffledList;
+        return;
+      }
+
+      this._shuffledList = fillIndices(this.length);
+      var oldIndex = null;
+      if (keepIndex) {
+        oldIndex = this._shuffledList.splice(this._index, 1)[0];
+        this._index = 0;
+      }
+
+      shuffle(this._shuffledList);
+
+      if (keepIndex) {
+        this._shuffledList.unshift(oldIndex);
+      }
+    },
+
+    /**
+     * Move the queue's current pointer to the next track.
+     *
+     * @param {Boolean} automatic True if this action is happening automatically
+     *        (i.e. when a track ends).
+     */
+    next: function(automatic = false) {
+      return this.advance(1, automatic);
+    },
+
+    /**
+     * Move the queue's current pointer to the previous track.
+     */
+    previous: function() {
+      this.advance(-1);
+    },
+
+    /**
+     * Move the queue's current pointer to another track.
+     *
+     * @param {Number} offset The offset to move by.
+     * @param {Boolean} automatic True if this action is happening automatically
+     *        (i.e. when a track ends).
+     */
+    advance: function(offset, automatic = false) {
+      var next;
+      var justShuffled = false;
+
+      // If the shuffle setting was changed after we last updated, shuffle
+      // (or unshuffle, as necessary).
+      if (this._shuffleTime < shuffleTime) {
+        this._shuffle(playbackSettings.shuffle);
+        justShuffled = true;
+      }
+
+      switch (PlaybackQueue.repeat) {
+      case Repeat.OFF:
+        next = Math.max(this._index + offset, 0);
+        break;
+      case Repeat.LIST:
+        next = posmod(this._index + offset, this.length);
+        var cycled = next !== this._index + offset;
+        // If we cycled around the list, reshuffle if necessary.
+        if (playbackSettings.shuffle && cycled && !justShuffled) {
+          this._shuffle(true);
+        }
+        break;
+      case Repeat.SONG:
```

Comment:
I wonder if it might be DRYer to combine this case with the `Repeat.OFF` case? Might not be as clear to read though.